### PR TITLE
`storage-plus`: Introduction and Basics

### PR DESCRIPTION
**Pull request by uint**
~~Review and merge #20 first!~~

### CODE REVIEW

File: src/pages/cw-storage-plus/basics.mdx [hunk 1]
```diff
@@ -0,0 +1,82 @@
+import { Callout } from "nextra/components";
+
+# Basics
+
+## Containers
+
+_cw-storage-plus_ provides a number of storage containers that can be used to
+store data on the blockchain. The fundamental ones are
+
+- [`Item`](containers/item),
```
Comment (aumetra) [inline src/pages/cw-storage-plus/basics.mdx]: I would maybe remove the commas at the end of the list

File: src/pages/cw-storage-plus/basics.mdx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+TODO: Research the key collisions in the warning below more! Then we can provide
+better advice.
+
+<Callout type="warning">
+  ...yet there's a small _but_. A `Map`'s prefix is length-prefixed. An `Item`
+  is saved without any sort of length-prefixing of the key. It is, in theory,
+  possible for an `Item`'s key to conflict with one of the keys generated by a
+  `Map` or `Deque`. In practice, this is unlikely unless you use very long
+  prefixes or start your `Item`'s key with the ␀ ASCII character. Probably don't
```
Comment (aumetra) [inline src/pages/cw-storage-plus/basics.mdx]: I think the ␀ looks cool but I had to search it up online to understand what it's supposed to be. Maybe we can say "null byte" instead?

File: src/pages/cw-storage-plus/basics.mdx [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+or [`Coin`](https://docs.rs/cosmwasm-std/2.0.3/cosmwasm_std/struct.Coin.html))
+often do as well.
+
+If you're writing a custom type and wish to send it across call or put it in
+storage, you'll have to derive these traits.
+
+<Callout>
+  *cw-storage-plus* uses
+  [*serde-json-wasm*](https://github.com/CosmWasm/serde-json-wasm) under the
+  hood. This provides determinism in some corners -
```
Comment (aumetra) [inline src/pages/cw-storage-plus/basics.mdx:81]: Small nit: it's also about space efficiency; `serde-json` is huge. But that doesn't have to be mentioned

File: src/pages/cw-storage-plus.md [hunk 4]
```diff
@@ -1,4 +1,41 @@
 # Introduction
 
-TODO: Describe what low-level interface `cw-storage-plus` builds on, and how it
-provides typed abstractions on top of it
+Not being able to persist data across calls would limit the utility of smart
+contracts. How could a smart contract
+
+- implement a token if it could not keep track of balances,
```
Comment (aumetra) [inline src/pages/cw-storage-plus.md]: Same here, I would remove the commas and the `, or` on the second entry

File: src/pages/cw-storage-plus.md [hunk 5]
```diff
@@ -1,4 +1,41 @@
...
+This is why a _CosmWasm_ smart contract has access to the storage facilities
+offered by the _Cosmos SDK_. These facilities are essentially a binary key-value
+store, with records stored on-chain.
+
+## Storage limits
+
+While developing smart contracts, it's important to remember on-chain storage
+is, as always, pricey. Conventionally, developers often draw the line at a
+"small logo image" (think a few KBs). If you need to store bigger things, it's
+likely time to consider off-chain storage (like IPFS or centralized storage).
```
Comment (aumetra) [inline src/pages/cw-storage-plus.md]: > (like IPFS or centralized storage)

This sounds a little wonky gramatically without a glue word.

```suggestion
likely time to consider off-chain storage (like IPFS or some centralized storage).
```

File: src/pages/cw-storage-plus.md [hunk 6]
```diff
@@ -1,4 +1,41 @@
...
+
+The smart contract framework itself (_cosmwasm-std_) provides a simple API for
+storing and retrieving data. If you're curious, you can check it out right
+[here](https://docs.rs/cosmwasm-std/2.0.3/cosmwasm_std/trait.Storage.html).
+
+This API is raw in that it exposes the **binary** key-value store. While you're
+free to use it directly, you're likely to find that finicky and error-prone.
+_cw-storage-plus_ is a library that builds on top of this API to
+
+- eliminate the need to manually serialize and deserialize data,
```
Comment (aumetra) [inline src/pages/cw-storage-plus.md]: Same comment about commas here

**Review by aumetra — COMMENTED** [review]
In general this looks really good!  
Just a general stylistical question, should we maybe wrap crate names in backticks (`; so inline code blocks) instead of wrapping them in stars (*).

Because I see italics for emphasising things, while backticks are more suited for technical things (such as *crate names* and similar things)

File: src/pages/cw-storage-plus/basics.mdx [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+An [`Item`](containers/item) is a simple container that stores a single value.
+The other two are a little more involved - they are collections capable of
+storing multiple values, and provide several methods to interact with them.
+
+## Keys and prefixes
+
+One task of a storage library like `cw-storage-plus` is to manage the namespace
+of keys provided by the blockchain.
+
+When constructing a container, you must provide a key of type `&'static str`.
```
Comment (dakom) [inline src/pages/cw-storage-plus/basics.mdx]: doesn't `new_dyn()` allow getting around this restriction?

File: src/pages/cw-storage-plus/basics.mdx [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+This usually means you'll be providing a string literal or some constant.
+
+In the case of an [`Item`](containers/item), the provided key is the exact key
+(a standard UTF-8 string) under which the value will be saved. In the case of
+collections, the provided key is really a prefix; the collection will use this
```
Comment (dakom) [inline src/pages/cw-storage-plus/basics.mdx]: maybe say that this is a "namespace"?

**Review by dakom — APPROVED** [review]
a couple small comments, and I agree with @aumetra 's - overall LGTM! :)

File: src/pages/cw-storage-plus/basics.mdx (hunk 1 above)
Comment (uint) [reply src/pages/cw-storage-plus/basics.mdx]: If that's how everyone's been doing lists, I'm happy to follow suit, sure. In general this isn't wrong AFAIK; it's just a question of "which style guide you're following".

**Comment by uint** [comment]
> In general this looks really good! Just a general stylistical question, should we maybe wrap crate names in backticks (`; so inline code blocks) instead of wrapping them in stars (*).
> 
> Because I see italics for emphasising things, while backticks are more suited for technical things (such as _crate names_ and similar things)

@aumetra I tend to think of italics as used for proper names (like *Newsweek* or *CosmWasm*), and inline code as mostly used for code, variable names, types, and the like. I guess for me crate names tend to fall in the "proper name" category, though I see how that's blurry. Slack poll?

**Comment by aumetra** [comment]
I think if you rebase the PR, that should pick up commits fixing that deploy problem

File: src/pages/cw-storage-plus/basics.mdx (hunk 8 above)
Comment (uint) [reply src/pages/cw-storage-plus/basics.mdx]: I would find "the provided key is really a namespace" a bit confusing here if that's what you mean. Are you worried the name of the parameter (`namespace`) in constructors will be confusing?

**Comment by uint** [comment]
Rebased!

File: src/pages/cw-storage-plus/basics.mdx (hunk 8 above)
Comment (dakom) [reply src/pages/cw-storage-plus/basics.mdx]: I'm not sure what exactly is a bit confusing - but vaguely that in collections, keys are not typically what you think of as the namespace/prefix, but rather the thing you pass to save/load/etc.

how about: "Collections are created with a namespace which acts as a prefix for all keys; the collection will use this..." ?

Comment (uint) [reply src/pages/cw-storage-plus/basics.mdx]: The problem I'm having is that AFAIK a namespace is a set of names, while (for us) a prefix is a thing that generates that set of names.

Would this work?

```md
In the case of an [`Item`](containers/item), the provided string is the exact key
(a standard UTF-8 string) under which the value will be saved. In the case of
collections, the provided string is a prefix (called `namespace` in code); the collection will use this
```

Comment (dakom) [reply src/pages/cw-storage-plus/basics.mdx]: yeah, that clears it up for me :)

**Comment by uint** [comment]
I think I applied all the suggestions. Thanks a lot.

I'm going to merge since I have approval, but if there's still something to fix, I'm happy to open a new PR.

**Pull request merged**